helios-client/src/main/java/com/spotify/helios/auth/CrtAuthProvider.java

```diff
@@ -0,0 +1,159 @@
+/*
+ * Copyright (c) 2015 Spotify AB.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.spotify.helios.auth;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Function;
+import com.google.common.base.Throwables;
+import com.google.common.util.concurrent.AsyncFunction;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+
+import com.spotify.crtauth.CrtAuthClient;
+import com.spotify.crtauth.exceptions.CrtAuthException;
+import com.spotify.crtauth.signer.Signer;
+import com.spotify.helios.client.HeliosRequest;
+import com.spotify.helios.client.RequestDispatcher;
+import com.spotify.helios.client.Response;
+
+import org.jetbrains.annotations.NotNull;
+
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.concurrent.atomic.AtomicReference;
+
+import javax.annotation.Nullable;
+
+import static com.google.common.base.Strings.isNullOrEmpty;
+
+public class CrtAuthProvider implements AuthProvider {
+
+  private static final String CRT_HEADER = "X-CHAP";
+  private static final String AUTH_URI = "https://helios/_auth";
+  private static final String CHALLENGE_PREFIX = "challenge";
+  private static final String TOKEN_PREFIX = "token";
+
+  private final RequestDispatcher dispatcher;
+  private final String username;
+  private final CrtAuthClient crtClient;
+  private final AtomicReference<String> tokenRef;
```

Comment:
the problem is if it's not final, I can't reference it inside the anonymous `Function` class. And if it is final, I can't change it. So I need a reference to a String instead of a String itself.